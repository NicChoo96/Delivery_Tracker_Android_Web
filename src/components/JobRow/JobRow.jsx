import React, { useRef, useState, useEffect, useCallback } from 'react';
import './JobRow.scss';
import {
  Button,
  Tooltip,
  Zoom,
  TableCell,
  TableRow,
} from '@material-ui/core';
import { history, copyToClipboard, approximateAddress, timestampConversion } from '../../utils';

/**
 * Renders one job row with the job data such as datetime, location, etc
 * @category UI Component
 * @subcategory JobTable
 * @class JobRow
 * @param {string} jobID Job ID
 * @param {Array<locationData>} location array coordinate and timestamp of contractor's location history
 * @param {string} company contractor's company
 * @param {number} phone contractor's phone number
 * @param {number} contractorName contractor's name
 * @param {number} dateTimeStart timestamp when the job starts
 * @param {number} dateTimeStart timestamp when the job ends
 * @return {ReactComponent}
 */
function JobRow({ jobID, location, company, phone, status, contractorName, dateTimeStart, dateTimeEnd }) {
  const statusColor = 'job-row__status-' + (status === 'Complete' ? 'completed' : 'inprogress');
  const jobIdRef = useRef();
  const [tooltipText, setTooltipText] = useState(jobID);
  const [locationName, setLocationName] = useState(' - ');

  const memoizedApproximatAddress = useCallback(() => approximateAddress(
    location[location.length - 1].lat,
    location[location.length - 1].lng
  ), [location]);

  useEffect(() => {
    if (location && location.length > 0) {
      memoizedApproximatAddress()
        .then(data => setLocationName(data))
        .catch(error => {
          setLocationName(' - ');
          console.error(error);
        });
    } else {
      setLocationName(' - ')
    }
  }, [memoizedApproximatAddress, location]);


  const copyIdHandler = (inputRef) => (event) => {
    event.preventDefault();
    event.stopPropagation();

    copyToClipboard(inputRef.current.innerHTML);
    setTooltipText('Copied');
    setTimeout(() => setTooltipText(jobID), 1000);
  };

  return (<TableRow key={jobID} className="job-row__row">
    <TableCell align="left">
      <Tooltip
        title={tooltipText || ''}
        TransitionComponent={Zoom}
        placement="right"
        arrow
      >
        <Button
          style={{ textTransform: 'none' }}
          onClick={copyIdHandler(jobIdRef)}
          className="job-row__id_btn"
          variant="text"
        >
          <span ref={jobIdRef} className="job-row__id_string">
            {jobID}
          </span>
        </Button>
      </Tooltip>
    </TableCell>
    <TableCell align="left"><span className="app__text-cursor">{locationName}</span></TableCell>
    <TableCell align="left"><span className="app__text-cursor job-row__cell-no-wrap">{timestampConversion(dateTimeStart)}</span></TableCell>
    <TableCell align="left"><span className="app__text-cursor job-row__cell-no-wrap">{(dateTimeEnd) ? timestampConversion(dateTimeEnd) : `-`}</span></TableCell>
    <TableCell align="left"><span className="app__text-cursor">{contractorName}</span></TableCell>
    <TableCell align="left"><span className="app__text-cursor">{company}</span></TableCell>
    <TableCell align="left"><span className="app__text-cursor job-row__cell-no-wrap">{phone}</span></TableCell>
    <TableCell align="center"><span className={statusColor + " app__text-cursor job-row__cell-no-wrap"}>{status}</span></TableCell>
    <TableCell align="left">
      <Button
        variant="contained"
        size="small"
        color="primary"
        onClick={() => history.push('/jobs/' + jobID)}
        disabled={locationName === (location && location.length > 0)}
      >
        View
          </Button>
    </TableCell>
  </TableRow>
  );
}

export default JobRow;
