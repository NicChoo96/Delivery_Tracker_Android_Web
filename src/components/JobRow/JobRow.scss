.job-row__status-completed,
.job-row__status-inprogress {
  border-radius: 5px;
  text-align: center;
  padding: 0.5em;
  white-space: nowrap;
}

.job-row__status-completed {
  background-color: #33eb91;
}

.job-row__status-inprogress {
  background-color: #ffef62;
}

.job-row__id_btn {
  max-width: 100%;
}

.job-row__id_string {
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
  max-width: 5em;
}

.job-row__cell-no-wrap {
  white-space: nowrap;
}

:nth-child(odd).job-row__row {
  background-color: #f3f3f3;
}

.job-row__row:hover {
  background-color: #ddffff;
}
