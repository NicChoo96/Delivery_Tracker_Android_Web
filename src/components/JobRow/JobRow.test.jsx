import React from 'react';
import ReactDOM from 'react-dom';
import '@testing-library/jest-dom/extend-expect';

import { cleanup } from '@testing-library/react';
import renderer from 'react-test-renderer';

import JobRow from './JobRow';

const testValues = {
  jobID: 'test id',
  location: [],
  company: 'imma real company',
  phone: '69696969',
  status: 'completed',
  contractorName: 'imma real contractor',
  dateTimeStart: '69',
  dateTimeEnd: '69',
};

afterEach(cleanup);

describe('JobRow', () => {
  test('renders without crashing', () => {
    const table = document.createElement('table');
    const tableBody = document.createElement('tbody');
    table.appendChild(tableBody);
    ReactDOM.render(<JobRow />, tableBody);
  });
//jobID, location, company, phone, status, contractorName, dateTimeStart, dateTimeEnd
  test('test values of component props', () => {
    const component = (
      <JobRow
        data-testid="jobrow__testprops"
        {...testValues}
      />
    );
    Object.keys(component).map((key) => {
      return expect(component.props.key).toBe(testValues.key);
    });
  });

  test('matches snapshot', () => {
    const tree = renderer.create(<JobRow {...testValues} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
