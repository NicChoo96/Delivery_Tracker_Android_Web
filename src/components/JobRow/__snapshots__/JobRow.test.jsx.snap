// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JobRow matches snapshot 1`] = `
<tr
  className="MuiTableRow-root job-row__row"
  role={null}
>
  <td
    aria-sort={null}
    className="MuiTableCell-root MuiTableCell-alignLeft"
  >
    <button
      aria-describedby={null}
      className="MuiButtonBase-root MuiButton-root MuiButton-text job-row__id_btn"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseOver={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      style={
        Object {
          "textTransform": "none",
        }
      }
      tabIndex={0}
      title="test id"
      type="button"
    >
      <span
        className="MuiButton-label"
      >
        <span
          className="job-row__id_string"
        >
          test id
        </span>
      </span>
      <span
        className="MuiTouchRipple-root"
      />
    </button>
  </td>
  <td
    aria-sort={null}
    className="MuiTableCell-root MuiTableCell-alignLeft"
  >
    <span
      className="app__text-cursor"
    >
       - 
    </span>
  </td>
  <td
    aria-sort={null}
    className="MuiTableCell-root MuiTableCell-alignLeft"
  >
    <span
      className="app__text-cursor job-row__cell-no-wrap"
    >
      1 Jan - 07:30 AM
    </span>
  </td>
  <td
    aria-sort={null}
    className="MuiTableCell-root MuiTableCell-alignLeft"
  >
    <span
      className="app__text-cursor job-row__cell-no-wrap"
    >
      1 Jan - 07:30 AM
    </span>
  </td>
  <td
    aria-sort={null}
    className="MuiTableCell-root MuiTableCell-alignLeft"
  >
    <span
      className="app__text-cursor"
    >
      imma real contractor
    </span>
  </td>
  <td
    aria-sort={null}
    className="MuiTableCell-root MuiTableCell-alignLeft"
  >
    <span
      className="app__text-cursor"
    >
      imma real company
    </span>
  </td>
  <td
    aria-sort={null}
    className="MuiTableCell-root MuiTableCell-alignLeft"
  >
    <span
      className="app__text-cursor job-row__cell-no-wrap"
    >
      69696969
    </span>
  </td>
  <td
    aria-sort={null}
    className="MuiTableCell-root MuiTableCell-alignCenter"
  >
    <span
      className="job-row__status-inprogress app__text-cursor job-row__cell-no-wrap"
    >
      completed
    </span>
  </td>
  <td
    aria-sort={null}
    className="MuiTableCell-root MuiTableCell-alignLeft"
  >
    <button
      className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-containedSizeSmall MuiButton-sizeSmall"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      <span
        className="MuiButton-label"
      >
        View
      </span>
      <span
        className="MuiTouchRipple-root"
      />
    </button>
  </td>
</tr>
`;
